### ISSUE TITLE
Value estimator training

### ISSUE THREAD

**Issue by Harry-mic** [idanshen/Value-Augmented-Sampling#1]
Hi! Thanks for your awesome work! 

I have some confusion on the training process of the Value estimator. As you mention in the paper's Appendix,C.1. VAS Training:
“In addition, we also collected a validation dataset by first generating a dataset of half-completed responses and, for each one, generating ten completions and averaging their reward. ” I wonder if this validation set format is only used for evaluation, or also in the training process. 

Because TD algorithm need several rewards from the environment in one trajectory to update, if one training QA only have one reward after the last token, the other tokens' rewards will be zero. I check the training code of TinyLlama-1B in this repo, and find the reward is only calculated once after the entire sentence.

If you introduce multi-reward in one trajectory, we will be very grateful if you'd open source this code !
Thanks again for your time!

**Comment by idanshen** [comment]
Thank you for the answers.

TD algorithms don't necessarily need multiple rewards in one trajectory and can work just fine with a single reward at the end of the trajectory. This is what we did for our experiments.

In the dense reward case, the reward is usually a vector the length of the trajectory. Therefore, you can just comment out lines 708-714 in vas_trainer.py

**Comment by Harry-mic** [comment]
Thanks for your careful reply! It solves most of my issues. By the way, I'm curious about how you use this validation set. Is it used to early stop the training of value estimator and what metric do you use ? Thanks again for your help!

**Comment by idanshen** [comment]
The validation set was mainly used for validation and analysis (see section 4.7 in our paper). We especially found it useful during our initial experiments as it gave us some metric about how good a training session is without needing a full decoding and rewarding of the test set.

**Comment by Harry-mic** [comment]
Thanks for your reply! I wonder how is the Validation MSE calculated, do you first generate a response using the value model, and then generate ten without value model, then calculate their loss?
